tools/yarn-cling/README.md

```diff
@@ -0,0 +1,78 @@
+# yarn-cling
+
+Generate an NPM shrinkwrap file from a yarn-managed monorepo.
+
+## Why do we need an NPM shrinkwrap file?
+
+When vending JavaScript applications that are installed via NPM, an
+`npm-shrinkwrap.json` is necessary to control the dependency tree of the
+installed application, ensuring that all the dependencies have the
+version that the application vendor expects.
+
+1. This prevents `npm install <application>` on the user's computer
+   from installing an untested combination of versions, one that may potentially
+   be broken. This *shouldn't* happen if everyone nicely keeps to semantic
+   versioning, but doing so relies on good intentions.
+
+2. Since most package's dependencies are written like `^1.2.0`, any application
+   in the NPM ecosystem can potentially be compromised by someone releasing a
+   minor or patch version of a library somewhere deep in the dependency tree
+   with malware in it. All subsequent `npm install <application>`s would happily
+   install the new version of the now compromised library.
+
+The only way around both of these issues is an `npm-shrinkwrap.json`, which will
+be respected by NPM on doing `npm install` (unlike `package-lock.json`, which
+won't).
+
+Note that yarn doesn't support shrinkwrapping at all. We can't help those
+people, but we can at least protect NPM users and tell people to use NPM to
+install our applications if they want to have some semblance of installation
+safety.
+
+## Okay fine. Why does this tool need to exist?
+
+There doesn't seem to be any existing tool that can generate the
+`npm-shrinkwrap.json` file from our monorepo.
+
+### What about 'npm shrinkwrap' ?
+
+There is the command `npm shrinkwrap`. From various Googles, this command
+varyingly used to accept arguments and not accept arguments. Its current
+incarnation on my NPM (6.11.3) does not accept arguments, and simply
+renames `package-lock.json => npm-shrinkwrap.json`. We don't have
+a `package-lock.json` (because we manage our monorepo completely using
+Yarn), so that obviously won't work.
+
+Nevertheless, if I run `npm shrinkwrap` a file IS generated. This file
+contains SOME version information, but doesn't contain package integrity
+checksums and breaks NPM when a subsequent `npm install` is run. NPM
```

Comment:
This seems strange. Does it have to do with our monorepo as well? Otherwise it implies this tool is just broken?
Running the tool locally I did get integrity checksums:

```json
{
  "name": "aws-cdk",
  "version": "0.0.0",
  "lockfileVersion": 1,
  "requires": true,
  "dependencies": {
    "@types/node": {
      "version": "10.17.18",
      "dev": true
    },
    "abbrev": {
      "version": "1.1.1",
      "resolved": "https://registry.npmjs.org/abbrev/-/abbrev-1.1.1.tgz",
      "integrity": "sha512-nne9/IiQ/hzIhY6pdDnbBtz7DjPTKrY00P/zvPSm5pOFkl6xuGrGnXn/VtTNNfNtAfZ9/1RtehkszU9qcTii0Q=="
    },
    "ansi-styles": {
      "version": "1.0.0",
      "resolved": "https://registry.npmjs.org/ansi-styles/-/ansi-styles-1.0.0.tgz",
      "integrity": "sha1-yxAt8cVvUSPquLZ817mAJ6AnkXg="
    },
    "ansicolors": {
      "version": "0.2.1",
      "resolved": "https://registry.npmjs.org/ansicolors/-/ansicolors-0.2.1.tgz",
      "integrity": "sha1-vgiVmQl7dKXJxKhKDNvNtivYeu8="
    },
    ...
}
```

Reply:
Ok - shrinkwrap only generates the correct file if you run `npm install` beforehand (which generates a package-lock.json file to be renamed). We can't do that because it will bypass our yarn.lock file and use potentially different deps. Resolving this discussion.